From 5beb5e7ca7718d39c8bbfd7dc658a9246a68b871 Mon Sep 17 00:00:00 2001
From: Bruce E. Robertson <bruce.e.robertson@intel.com>
Date: Mon, 3 Dec 2012 13:01:28 -0800
Subject: ADBD: set sent file permissions per android fs config table.

This fix allows the standard CTS installation to succeed while
preserving the previous fix's intent to allow init to operate
properly.

The first CTS APK installation failed when adbd, the adb daemon, on
the target invoked pm (package manager). "pm" runs as another user
than adbd and could not open the APK because adbd had cleared the
"other" read permission.

For: AXIA-813
Change-Id: If27a628072f51ffac78db7b5448670ea4666575f
Signed-off-by: Bruce E. Robertson <bruce.e.robertson@intel.com>

Conflicts:
	adb/file_sync_service.c
---
 adb/Android.mk          |    2 ++
 adb/file_sync_service.c |   10 +++++++---
 2 files changed, 9 insertions(+), 3 deletions(-)

diff --git a/adb/Android.mk b/adb/Android.mk
index a803978..62d7cf6 100644
--- a/adb/Android.mk
+++ b/adb/Android.mk
@@ -119,6 +119,8 @@ LOCAL_SRC_FILES := \
 	log_service.c \
 	utils.c
 
+LOCAL_C_INCLUDES += $(LOCAL_PATH)/../include/private
+
 LOCAL_CFLAGS := -O2 -g -DADB_HOST=0 -Wall -Wno-unused-parameter
 LOCAL_CFLAGS += -D_XOPEN_SOURCE -D_GNU_SOURCE
 
diff --git a/adb/file_sync_service.c b/adb/file_sync_service.c
index d3e841b..429b58f 100644
--- a/adb/file_sync_service.c
+++ b/adb/file_sync_service.c
@@ -30,6 +30,7 @@
 #define TRACE_TAG  TRACE_SYNC
 #include "adb.h"
 #include "file_sync_service.h"
+#include "android_filesystem_config.h"
 
 static int mkdirs(char *name)
 {
@@ -306,10 +307,13 @@ static int do_send(int s, char *path, char *buffer)
 #else
     {
 #endif
-        /* copy user permission bits to "group" and "other" permissions */
-        mode |= ((mode >> 3) & 0070);
-        mode |= ((mode >> 3) & 0007);
+        unsigned uid, gid;
 
+        // strip all leading '/'(s)
+        for(tmp = path; *tmp == '/'; tmp++)
+            ;
+
+        fs_config(tmp, 0, &uid, &gid, &mode);
         ret = handle_send_file(s, path, mode, buffer);
     }
 
-- 
1.7.4.1

